bindings/cs/rl.net/RlLogger.cs

```diff
@@ -0,0 +1,337 @@
+        {
+            return this.TryLog(contextJson, response, apiStatus);
+        }
+
+        public bool TryLog(string contextJson, DecisionResponse response, ApiStatus apiStatus = null)
+        {
+            int result = LoggerLogCbInteraction(this.DangerousGetHandle(), contextJson, response.DangerousGetHandle(), apiStatus.ToNativeHandleOrNullptrDangerous());
+            return result == NativeMethods.SuccessStatus;
+        }
+
+        public void Log(string contextJson, DecisionResponse response)
+        {
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(contextJson, response, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        public bool Log(string eventId, float outcome, ApiStatus apiStatus = null)
+        {
+            return this.TryLog(eventId, outcome, apiStatus);
+        }
+
+        public bool TryLog(string eventId, float outcome, ApiStatus apiStatus = null)
+        {
+            int result = LoggerLogOutcomeF(this.DangerousGetHandle(), eventId, outcome, apiStatus.ToNativeHandleOrNullptrDangerous());
+            return result == NativeMethods.SuccessStatus;
+        }
+
+        public void Log(string eventId, float outcome)
+        {
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(eventId, outcome, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        public bool Log(string eventId, string outcome, ApiStatus apiStatus = null)
+        {
+            return this.TryLog(eventId, outcome, apiStatus);
+        }
+
+        public bool TryLog(string eventId, string outcome, ApiStatus apiStatus = null)
+        {
+            int result = LoggerLogOutcomeJson(this.DangerousGetHandle(), eventId, outcome, apiStatus.ToNativeHandleOrNullptrDangerous());
+            return result == NativeMethods.SuccessStatus;
+        }
+
+        public void Log(string eventId, string outcome)
+        {
+            using (ApiStatus apiStatus = new ApiStatus())
+            {
+                if (!this.TryLog(eventId, outcome, apiStatus))
+                {
+                    throw new RLException(apiStatus);
+                }
+            }
+        }
+
+        private event EventHandler<ApiStatus> BackgroundErrorInternal;
+
+        // This event is thread-safe, because we do not hook/unhook the event in user-schedulable code anymore.
+        public event EventHandler<ApiStatus> BackgroundError
+        {
+            add
+            {
+                this.BackgroundErrorInternal += value;
+            }
+            remove
+            {
+                this.BackgroundErrorInternal -= value;
+            }
+        }
+
+        private event EventHandler<TraceLogEventArgs> OnTraceLoggerEventInternal;
```

Comment:
Why does this exists?

Reply:
There is no reason for this to be an event. Make it a delegate.

Reply:
I am not sure I follow what you mean by changing this to a delegate? This is used to enable native=>managed transition hooking/unhooking inside of the add/remove handlers for TraceLoggerEvent.

Reply:
Right now this pattern exactly replicating the one from LiveModel. If there are remaining issues that has to be fixed in both places, let's do it in separate PR.